### ISSUE TITLE
CI: Improve coverage for verifier complexity

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#5058]
### Background

The datapath code is generated based upon a set of `#define` statements in header files. When Cilium is deployed, the agent generates these macros in the header files, and invokes the compiler to compile a specific combination of options. When the customised datapath code has been generated, the agent loads this code into the kernel. If the generated code is too complex, or the kernel's verifier cannot successfully prune similar states, then it may reject the program.

#### Verifier testing

In the `bpf/` directory, we include hand-written copies of the header files that would normally be generated by the agent at runtime. These are used to perform compile-testing of the `bpf/` code in the tree. Furthermore, the `test/bpf/verifier-test.sh` script and `test/runtime/verifier.go` test make use of these to attempt to load the most complex BPF program that we might generate into the kernel during a CI run, so that (in theory) we can catch verifier compilation errors before they make it into the tree, into a release, and into user environments (where configuring too many options could lead to failure to generate programs for an endpoint, and hence connectivity loss for such affected endpoints).

### Problem Statement

With many different options that can be enabled or disabled in the header files, it is very difficult to predict how different features may affect the complexity. Originally I had thought that if all features are enabled and turned up to the maximum, then this would cause the complexity limit to hit its maximum. However, https://github.com/cilium/cilium/pull/5052 proved that this is not the case; in that PR, when disabling `POLICY_INGRESS` / `POLICY_EGRESS`, enabling `DEBUG`, disabling `DEBUG_LB`, and setting `MONITOR_AGGREGATION_LEVEL` to 0 rather than 5, the complexity limit was exceeded. Swapping any of these options back to their defaults would cause the complexity to reduce under the limit.

### Proposal

* [x] Create a nightly test which runs the `test/bpf/verifier-test.sh` script, https://github.com/cilium/cilium/pull/12658.
* [ ] Enumerate the set of possible options for the datapath
  * Include definitions like the ones specified above
  * Don't include different definitions of stuff like `LXC_ID`, map names, etc.
  * This likely would involve abstracting the datapath configuration generation logic in `pkg/datapath/config` and writing a standalone tool to generate the `ep_config.h` from a random seed.
* [ ] Extend this nightly test to randomly generate different options for the datapath, build the `bpf/` directory, and run `verifier-test.sh`.
  * [x] Change `bpf/Makefile` to be able to compile a single bpf_xxx target at a time while passing `MAX_XXX_OPTIONS` as an environment variable. https://github.com/cilium/cilium/pull/17364
  * [ ] Change `test/bpf/verifier-test.sh` to be able to load a single bpf_xxx program at a time. https://github.com/cilium/cilium/pull/17408
  * [ ] Change `test/k8sT/Verifier.go` to go over the list of hardcoded configurations for each bpf_xxx target, compile the bpf_xxx program, and load it. https://github.com/cilium/cilium/pull/17470

This could be sufficient to run on the oldest possible kernel that we support (whatever is in CI) as it is unlikely that a newer kernel will analyse a program to be more complex than the assessment of an older kernel.

**Comment by joestringer** [comment]
An alternative to this proposal might be to reduce (eliminate?) the number of possible build-time options in the datapath. If all of the code is always the same, then one run of the verifier would be enough. AFAIK there are no current plans to do this.

**Comment by pchaigno** [comment]
> Extend this nightly test to randomly generate different options for the datapath

I agree with the general outlined proposed, except maybe for that one :point_up: The goal of such tests is to give us reassurances that we don't have complexity regressions. If our test only checks a randomly-generated option combination (or even a set of randomly-generated option combinations), it won't give us the reassurance we need to merge PRs and make releases.

I think it's possible to generate a static set of all option combinations we want to test and validate all of them on each test run. The test will be longer and maybe only run on master by default, but I think it's worth it if it tells us "it's highly unlikely we introduced a new complexity regression".

### New Proposal

So the proposal would then become:
- [ ] Enumerate the set of possible options for the datapath.
    - I think it's possible to extract the sets of options to validate from our C code automatically. We would of course not generate all possible combinations, but heuristically try to generate all local maxima combinations (see below).
    - Despite the above, we may also need to include hand-selected option combinations, known to cause complexity issues.
- [ ] Allow K8sVerfiier to run with those sets of datapath options.
    - [ ] Adapt `bpf/Makefile` to take a set of datapath options as input. https://github.com/cilium/cilium/issues/16414 would be useful here.
    - [ ] Extend K8sVerifier to be able to run more than one set of datapath options.

These two parts can be developed in parallel by testing the second part with hand-selected, hardcoded sets of datapath options.

### Local Maxima Combinations

We can define our ideal result as an abstract, N-dimensional function `complexity(X) -> y`, where N is the number of different options, `X` is the values of those options (typically 0 or 1), and `y` is the complexity. Tracing that function would require us to compile and load programs for all possible option combination, which would be prohibitively expensive.

Fortunately, we don't need all values of `complexity`; we only need to know that is never goes above a threshold. Or maybe we are interested in all high values because we want to monitor their evolution. But we sure don't care about e.g. any `X` that generates a `y=100`.

What our current K8sVerifier test attempts to do is guess the `Xmax` that gives the maximum of `complexity` and measure `complexity(Xmax)`. That doesn't work really well because it's hard to guess this `Xmax`* and it's value can change rapidly even with small code changes. Instead, I think we should try to monitor all `Xi` that are likely to be local maxima. We're more likely to correctly guess the maximum if we track (almost) all local maxima.

We can guess those intuitively from our C code. In the code below, we may want to test `X={ENABLE_A=1,ENABLE_C=1}` and `X={ENABLE_B=1}`, but we don't care about `X={ENABLE_A=1,ENABLE_B=1}` and `X={ENABLE_B=1,ENABLE_C=1}`. More importantly for our heuristic, we can probably also skip testing `X={ENABLE_A=1}` because `X={ENABLE_A=1,ENABLE_C=1}` likely has a higher value.

```c
#ifdef ENABLE_A
  ... code ...
# ifdef ENABLE_C
  ... code ...
# endif
#elif ENABLE_B
  ... code ...
#endif
```

Here our local maxima are `X={ENABLE_A=1,ENABLE_C=1}` and `X={ENABLE_B=1}`.

\* No, it's not `{25,12,2021}`.

**Comment by pchaigno** [comment]
After our ah-hoc call to discuss this and @kkourt's first estimation of the number of configurations we'd have to cover, I started making a small script to list all possible configurations while taking constraints between options into account. Commit https://github.com/pchaigno/cilium/commit/c672cd9f3fff5257bc29b3420eb7e26f255ffed5 has an explanation of the script and how it works. Commit https://github.com/pchaigno/cilium/commit/6175c7210333c1d9055422efe30ca669104cc855 contains the raw results.

We end up with ~28800 configurations to cover once we eliminate those very unlikely to have complexity issues (e.g., only `ENABLE_IPV4`). That's still too much to test regularly (8h x the number of seconds it takes to validate one configuration).

We discussed next steps with @kkourt:
- I will focus on enabling K8sVerifier to test more than one configuration. We need this anyway and we could always start with a couple of hardcoded configurations. That should ensure we make some progress on this item in the v1.11 cycle. This corresponds to the last item in the TODO list of this issue.
- @kkourt will dig into the list of generated configurations and measure the complexity for a subset or all of them. We're hoping this gives us a better idea of what we're dealing with and maybe helps us find new ways to reduce the search space. This will require a few changes and scripts:
    - Change `bpf/Makefile` to be able to compile a single bpf_xxx target at a time while passing MAX_XXX_OPTIONS as an environment variable.
    - Change `test/bpf/verifier-test.sh` to be able to load a single bpf_xxx program at a time. Some similar changes to `test/bpf/check-complexity.sh` may be needed.
    - Write a script to go through each (or a random subset of) configurations from https://github.com/pchaigno/cilium/commit/6175c7210333c1d9055422efe30ca669104cc855 and compile and load bpf_lxc/bpf_host with the above Makefile and Bash script.

We both need the changes to `bpf/Makefile` and `test/bpf/verifier-test.sh`, so we may have to sync around that :-)

**Comment by christarazi** [comment]
@pchaigno Is this done?

**Comment by pchaigno** [comment]
@christarazi No. We still rely on a very limited (2-3) set of datapath configurations in our complexity tests. They can be seen at `bpf/complexity-tests/`.
